other-cel/check-node-for-cve-2022-0185/check-node-for-cve-2022-0185.yaml

```diff
@@ -0,0 +1,34 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-kernel
+  annotations:
+    policies.kyverno.io/title: Check Node for CVE-2022-0185 in CEL expressions
+    policies.kyverno.io/category: Other in CEL 
+    policies.kyverno.io/severity: high
+    kyverno.io/kyverno-version: 1.11.0
+    policies.kyverno.io/minversion: 1.11.0
+    kyverno.io/kubernetes-version: "1.26-1.27"
+    policies.kyverno.io/subject: Node
+    policies.kyverno.io/description: >-
+      Linux CVE-2022-0185 can allow a container escape in Kubernetes if left unpatched.
+      The affected Linux kernel versions, at this time, are 5.10.84-1 and 5.15.5-2.
+      For more information, refer to https://security-tracker.debian.org/tracker/CVE-2022-0185.
+      This policy runs in background mode and flags an entry in the ClusterPolicyReport
+      if any Node is reporting one of the affected kernel versions.
+spec:
+  validationFailureAction: Audit
+  background: true
+  rules:
+  - name: kernel-validate
+    match:
+      any:
+      - resources:
+          kinds:
+            - Node
+    validate:
+      cel:
+        expressions:
+          - expression: "!(object.status.nodeInfo.kernelVersion in ['5.10.84-1', '5.15.5-2'])"
```

Comment:
Have you tried this policy? I am not sure if we could access the `status` field.

Reply:
I'm running a single node kind cluster. This is the node's configuration:
```
$ k get nodes -o jsonpath='{.items[*].status.nodeInfo}' | jq .
{
  "architecture": "amd64",
  "bootID": "684deeaf-72c0-463c-9a17-c24cc3d15391",
  "containerRuntimeVersion": "containerd://1.7.13",
  "kernelVersion": "6.5.0-1015-aws",
  "kubeProxyVersion": "v1.29.2",
  "kubeletVersion": "v1.29.2",
  "machineID": "1098f0df08574920a984b8f198a89076",
  "operatingSystem": "linux",
  "osImage": "Debian GNU/Linux 12 (bookworm)",
  "systemUUID": "35178929-44c5-4ff4-bf17-8e84eb1f3f48"
}
```

After applying the policy, this is the generated VAP:
```
$ k get validatingadmissionpolicy -o yaml
apiVersion: v1
items:
- apiVersion: admissionregistration.k8s.io/v1beta1
  kind: ValidatingAdmissionPolicy
  metadata:
    creationTimestamp: "2024-03-26T09:55:13Z"
    generation: 1
    labels:
      app.kubernetes.io/managed-by: kyverno
    name: check-kernel
    ownerReferences:
    - apiVersion: kyverno.io/v1
      kind: ClusterPolicy
      name: check-kernel
      uid: ba1a0d27-5466-4686-b0da-a9efa6f4209b
    resourceVersion: "2251"
    uid: d5f8d499-8b1e-46ca-b069-171136d9307a
  spec:
    failurePolicy: Fail
    matchConstraints:
      matchPolicy: Equivalent
      namespaceSelector: {}
      objectSelector: {}
      resourceRules:
      - apiGroups:
        - ""
        apiVersions:
        - v1
        operations:
        - CREATE
        - UPDATE
        resources:
        - nodes
        scope: '*'
    validations:
    - expression: '!(object.status.nodeInfo.kernelVersion in [''5.10.84-1'', ''5.15.5-2''])'
      message: Kernel is vulnerable to CVE-2022-0185.
  status:
    observedGeneration: 1
    typeChecking: {}
kind: List
metadata:
  resourceVersion: ""
```
This is the generated report we get:
```
$ k get cpolr -o yaml
apiVersion: v1
items:
- apiVersion: wgpolicyk8s.io/v1alpha2
  kind: ClusterPolicyReport
  metadata:
    creationTimestamp: "2024-03-26T09:55:24Z"
    generation: 1
    labels:
      app.kubernetes.io/managed-by: kyverno
    name: fe7942a7-da9b-4390-aaa6-a869a4accb13
    ownerReferences:
    - apiVersion: v1
      kind: Node
      name: vap-cluster-control-plane
      uid: fe7942a7-da9b-4390-aaa6-a869a4accb13
    resourceVersion: "2292"
    uid: fd5b66b9-5aa3-4b69-abc4-ce7f4d1744ac
  results:
  - category: Other in CEL
    message: Validation rule 'kernel-validate' passed.
    policy: check-kernel
    result: pass
    rule: kernel-validate
    scored: true
    severity: high
    source: kyverno
    timestamp:
      nanos: 0
      seconds: 1711446914
  scope:
    apiVersion: v1
    kind: Node
    name: vap-cluster-control-plane
    uid: fe7942a7-da9b-4390-aaa6-a869a4accb13
  summary:
    error: 0
    fail: 0
    pass: 1
    skip: 0
    warn: 0
kind: List
metadata:
  resourceVersion: ""
```
We are able to access the status field here.

Reply:
Great!